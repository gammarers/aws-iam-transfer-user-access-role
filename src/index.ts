import { Duration } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';


/**
 * @TODO: Not yet supported Omit
 * https://github.com/aws/jsii/issues/4468
 * type omitKeys = 'eventPattern';
 * export interface IamRDSMonitoringRoleProps extends Omit<iam.RoleProps, 'eventPattern'> {}
 */

interface IamOptionalRoleProps {
  readonly roleName?: string;
  readonly managedPolicies?: iam.IManagedPolicy[];
  readonly permissionsBoundary?: iam.IManagedPolicy;
  readonly maxSessionDuration?: Duration;
  readonly additionalInlinePolicies?: {
    [name: string]: iam.PolicyDocument;
  };
  readonly path?: string;
  readonly description?: string;
}

export interface IamTransferUserAccessRoleProps extends IamOptionalRoleProps {
  readonly bucket: s3.IBucket;
  readonly allowDir?: string;
}

export class IamTransferUserAccessRole extends iam.Role {
  constructor(scope: Construct, id: string, props: IamTransferUserAccessRoleProps) {

    super(scope, id, {
      roleName: props.roleName,
      managedPolicies: props.managedPolicies,
      permissionsBoundary: props.permissionsBoundary,
      maxSessionDuration: props.maxSessionDuration,
      path: props.path,
      description: (() => {
        return props.description ?? 'sftp user access role';
      })(),
      assumedBy: new iam.ServicePrincipal('transfer.amazonaws.com'),
      inlinePolicies: ((): { [name: string]: iam.PolicyDocument } => {
        const policies = {
          ['allow-bucket-access-policy']: new iam.PolicyDocument({
            statements: [
              new iam.PolicyStatement({
                sid: 'AllowListingOfUserFolder',
                effect: iam.Effect.ALLOW,
                actions: [
                  's3:ListBucket',
                  's3:GetBucketLocation',
                ],
                resources: [
                  props.bucket.bucketArn,
                ],
              }),
              new iam.PolicyStatement({
                sid: 'HomeDirObjectAccess',
                effect: iam.Effect.ALLOW,
                actions: [
                  's3:PutObject',
                  's3:PutObjectACL',
                  's3:GetObject',
                  's3:GetObjectACL',
                  's3:GetObjectVersion',
                  's3:DeleteObject',
                  's3:DeleteObjectVersion',
                ],
                resources: [
                  (() => {
                    if (props.allowDir) {
                      if (props.allowDir.charAt(props.allowDir.length - 1) === '/') {
                        // TODO: error when dir last char
                      }
                      if (props.allowDir.charAt(0) === '/') {
                        // TODO: error when dir first char
                      }
                      return props.bucket.bucketArn + '/' + props.allowDir + '/*';
                    }
                    return props.bucket.bucketArn + '/*';
                  })(),
                ],
              }),
            ],
          }),
        };
        return { ...policies, ...(props.additionalInlinePolicies || {}) };
      })(),
    });
  }
}
