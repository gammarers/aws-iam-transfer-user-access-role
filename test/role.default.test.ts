import { App, Stack } from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { IamTransferUserAccessRole } from '../src';

describe('IamTransferUserAccessRole default Testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  const bucket = new s3.Bucket(stack, 'TestBicket');

  const role = new IamTransferUserAccessRole(stack, 'IamTransferUserAccessRole', {
    bucket,
  });

  it('Is IAM Role', async () => {
    expect(role).toBeInstanceOf(iam.Role);
  });

  const template = Template.fromStack(stack);

  it('Should match iam role.', async () => {
    template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
      Description: 'sftp user access role',
      AssumeRolePolicyDocument: Match.objectEquals({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Principal: {
              Service: 'transfer.amazonaws.com',
            },
          },
        ],
      }),
      Policies: Match.arrayEquals([
        Match.objectEquals({
          PolicyName: 'allow-bucket-access-policy',
          PolicyDocument: Match.objectEquals({
            Version: '2012-10-17',
            Statement: Match.arrayEquals([
              Match.objectEquals({
                Sid: 'AllowListingOfUserFolder',
                Effect: 'Allow',
                Action: [
                  's3:ListBucket',
                  's3:GetBucketLocation',
                ],
                Resource: {
                  'Fn::GetAtt': [
                    Match.stringLikeRegexp('TestBicket'),
                    'Arn',
                  ],
                },
              }),
              Match.objectEquals({
                Sid: 'HomeDirObjectAccess',
                Effect: 'Allow',
                Action: [
                  's3:PutObject',
                  's3:PutObjectACL',
                  's3:GetObject',
                  's3:GetObjectACL',
                  's3:GetObjectVersion',
                  's3:DeleteObject',
                  's3:DeleteObjectVersion',
                ],
                Resource: {
                  'Fn::Join': [
                    '',
                    [
                      {
                        'Fn::GetAtt': [
                          Match.stringLikeRegexp('TestBicket'),
                          'Arn',
                        ],
                      },
                      '/*',
                    ],
                  ],
                },
              }),
            ]),
          }),
        }),
      ]),
    }));
  });

  it('Should match snapshot.', async () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

});