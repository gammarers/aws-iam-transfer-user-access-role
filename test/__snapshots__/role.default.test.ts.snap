// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IamTransferUserAccessRole default Testing Should match snapshot. 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "IamTransferUserAccessRole29572FE1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "transfer.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "sftp user access role",
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "s3:ListBucket",
                    "s3:GetBucketLocation",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "TestBicketC66E8595",
                      "Arn",
                    ],
                  },
                  "Sid": "AllowListingOfUserFolder",
                },
                {
                  "Action": [
                    "s3:PutObject",
                    "s3:PutObjectACL",
                    "s3:GetObject",
                    "s3:GetObjectACL",
                    "s3:GetObjectVersion",
                    "s3:DeleteObject",
                    "s3:DeleteObjectVersion",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        {
                          "Fn::GetAtt": [
                            "TestBicketC66E8595",
                            "Arn",
                          ],
                        },
                        "/*",
                      ],
                    ],
                  },
                  "Sid": "HomeDirObjectAccess",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "allow-bucket-access-policy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestBicketC66E8595": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
